docs: unstructured log parsing ADR (#2335)
kyma-project/telemetry-manager: docs/contributor/arch/023-unstructured-log-parsing.md

# 23. Unstructured Log Parsing

Date: 2025-07-30

## Status

Proposed

## Context

The current Fluent Bit-based LogPipeline custom resources offer minimal log parsing capabilities. However, with the ongoing migration toward OpenTelemetry (OTel)-based logging, we aim to maintain feature parity while improving the user experience and flexibility of configuration.

## Fluent Bit Log Parsing Overview

Fluent Bit supports various parsing mechanisms:
* Regex-based Parsing: Using the LogParser CR, applied to logs of specifically annotated workloads.
* Multiline Parsing: Through custom multiline filter.
* Built-in Parsers: For common log formats like JSON, Nginx, Apache, etc.

Limitations of Fluent Bit Approach
* Requires manual annotation of each pod to enable parsing.
* Limited support for mixed log formats within the same workload (needs further validation).
* Uses inconsistent APIs for different parsing types, which can confuse users.

## OTel Log Parsing: Requirements & Expectations

The OTel-based solution should fulfill the following requirements:

### Parser Definition

> Since LogPipeline have output (backend) affinity and parsers have input (workload) affinity, parsers must not be defined in the LogPipeline. Ideally, they should be defined in a separate resource. However, for now, we want to avoid introducing a new API and define parsers in the Telemetry CR.

Reviewer: ```suggestion
Because LogPipeline resources have output (backend) affinity and parsers have input (workload) affinity, parsers must not be defined in the LogPipeline. Ideally, they should be defined in a separate resource. However, for now, we want to avoid introducing a new API, and thus define parsers in the Telemetry CR.
```

### Workload Selection
Parsers must be tightly bound to a specific workload to ensure consistent application regardless of which pipeline is used.
It should be possible to bind parsers to workloads in a flexible way, allowing for:

>  * Broad targeting via namespace selectors

Reviewer: ```suggestion
 * Broad targeting with namespace selectors
```

 * Optional label-based selectors for advanced use cases

### Parser Types
We should support two main types of parsers:
 * Regex-based parsers
 * Multiline parsers
These can be used independently or in sequence, depending on the log structure.

### Built-in Parser Presets
We should support presets for common formats (Java, Python, Nginx) to reduce duplication and improve UX.

# Decision

Here is a proposed API:
```yaml
apiVersion: operator.kyma-project.io/v1alpha1
kind: Telemetry
metadata:
  name: default
  namespace: kyma-system
spec:
  log:
    parsers:
      - name: backend-parser
        namespaceSelector:
          name: backend-namespace
          # nameRegex: backend-namespace-.*
        podSelector:
          name: buttercup-app
          nameRegex: buttercup-app-.*
        containerSelector:
          name: server
          # nameRegex: server-.*
        multiline:
          type: builtin
          builtin: java
          # custom:
          #   firstEntryRegex: "^[^\\s]"
        regex:
          type: custom
          # builtin: nginx
          custom: "^Host=(?P<host>[^,]+), Type=(?P<type>.*)$"

```

## Tecnical Considerations

### Selector Support

> Parsers should be applicable via:

Reviewer: ```suggestion
Parsers should be applicable using the following options:
```
